File: tools/testing/connection/nodejs-client/gemini.mjs
#!/usr/bin/env node
/**
 * Gemini API (OpenAI Compatible) - Node.js Client Integration Test
 *
 * Authentication: Bearer Token
 * Base URL: https://generativelanguage.googleapis.com/v1beta/openai
 * Health Endpoint: GET /models
 *
 * Uses internal packages:
 *   - @internal/fetch-proxy-dispatcher: Environment-aware proxy configuration
 *   - @internal/fetch-client: HTTP client with auth support
 *   - @internal/provider_api_getters: API key resolution
 *   - @internal/app-static-config-yaml: YAML configuration loading
 */
import { fileURLToPath } from 'url';
import { dirname, resolve } from 'path';

// ============================================================================
// Project Setup
// ============================================================================
const __dirname = dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = resolve(__dirname, '..', '..', '..', '..');

// Load static config
const { loadYamlConfig, config: staticConfig } = await import(
  resolve(PROJECT_ROOT, 'packages_mjs', 'app-static-config-yaml', 'src', 'index.mjs')
);
const configDir = resolve(PROJECT_ROOT, 'common', 'config');
await loadYamlConfig({ configDir });

// Import internal packages
const { getProxyDispatcher } = await import(
  resolve(PROJECT_ROOT, 'packages_mjs', 'fetch-proxy-dispatcher', 'src', 'index.mts')
);
const { createClient } = await import(
  resolve(PROJECT_ROOT, 'packages_mjs', 'fetch-client', 'src', 'index.mts')
);
const { GeminiOpenAIApiToken, ProviderHealthChecker } = await import(
  resolve(PROJECT_ROOT, 'packages_mjs', 'provider_api_getters', 'src', 'index.mjs')
);

// ============================================================================
// Configuration - Exposed for debugging
// ============================================================================
const provider = new GeminiOpenAIApiToken(staticConfig);
const apiKeyResult = provider.getApiKey();

const CONFIG = {
  // From provider_api_getters
  GEMINI_API_KEY: apiKeyResult.apiKey,
  AUTH_TYPE: apiKeyResult.authType,

  // Base URL (from provider or override)
  BASE_URL: provider.getBaseUrl() || 'https://generativelanguage.googleapis.com/v1beta/openai',

  // Dispatcher (from fetch-proxy-dispatcher)
  DISPATCHER: getProxyDispatcher(),

  // Proxy Configuration (set to override YAML/environment config)
  PROXY: process.env.HTTPS_PROXY || process.env.HTTP_PROXY || undefined,

  // SSL/TLS Configuration (runtime override, or undefined to use YAML config)
  SSL_VERIFY: false,  // Set to undefined to use YAML config

  // Debug
  DEBUG: !['false', '0'].includes((process.env.DEBUG || '').toLowerCase()),
};

// ============================================================================
// Health Check
// ============================================================================
async function healthCheck() {
  console.log('\n=== Gemini Health Check (ProviderHealthChecker) ===\n');

  const checker = new ProviderHealthChecker(staticConfig);
  const result = await checker.check('gemini');

  console.log(`Status: ${result.status}`);
  if (result.latency_ms) console.log(`Latency: ${result.latency_ms.toFixed(2)}ms`);
  if (result.message) console.log(`Message: ${result.message}`);
  if (result.error) console.log(`Error: ${result.error}`);

  return { success: result.status === 'connected', result };
}

// ============================================================================
// Client Factory
// ============================================================================
function createGeminiClient() {
  return createClient({
    baseUrl: CONFIG.BASE_URL,
    dispatcher: CONFIG.DISPATCHER,
    auth: {
      type: 'bearer',
      rawApiKey: CONFIG.GEMINI_API_KEY,
    },
    proxy: CONFIG.PROXY,
    verify: CONFIG.SSL_VERIFY,
  });
}

// ============================================================================
// Sample API Calls using fetch-client
// ============================================================================
async function chatCompletion(messages, model = 'gemini-1.5-flash') {
  console.log(`\n=== Chat Completion (${model}) ===\n`);

  const client = createGeminiClient();

  try {
    const response = await client.post('/chat/completions', {
      json: { model, messages },
    });

    console.log(`Status: ${response.status}`);
    if (response.ok && response.data?.choices?.[0]?.message?.content) {
      console.log('Response:', response.data.choices[0].message.content);
    } else {
      console.log('Response:', JSON.stringify(response.data, null, 2));
    }

    return { success: response.ok, data: response.data };
  } finally {
    await client.close();
  }
}

async function streamChatCompletion(messages, model = 'gemini-1.5-flash') {
  console.log(`\n=== Streaming Chat Completion (${model}) ===\n`);

  const client = createGeminiClient();

  let fullContent = '';
  try {
    for await (const event of client.stream('/chat/completions', {
      method: 'POST',
      json: { model, messages, stream: true },
    })) {
      if (event.data === '[DONE]') continue;
      try {
        const data = JSON.parse(event.data);
        const content = data.choices?.[0]?.delta?.content || '';
        fullContent += content;
        process.stdout.write(content);
      } catch {
        // Ignore parse errors
      }
    }
    console.log('\n');
    return { success: true, content: fullContent };
  } finally {
    await client.close();
  }
}

async function createEmbedding(input, model = 'text-embedding-004') {
  console.log(`\n=== Create Embedding (${model}) ===\n`);

  const client = createGeminiClient();

  try {
    const response = await client.post('/embeddings', {
      json: { model, input },
    });

    console.log(`Status: ${response.status}`);
    if (response.ok && response.data?.data?.[0]?.embedding) {
      const embedding = response.data.data[0].embedding;
      console.log(`Embedding dimensions: ${embedding.length}`);
      console.log(`First 5 values: ${embedding.slice(0, 5).join(', ')}`);
    } else {
      console.log('Response:', JSON.stringify(response.data, null, 2));
    }

    return { success: response.ok, data: response.data };
  } finally {
    await client.close();
  }
}

// ============================================================================
// Run Tests
// ============================================================================
async function main() {
  console.log('Gemini API Connection Test (Node.js Client Integration)');
  console.log('='.repeat(55));
  console.log(`Base URL: ${CONFIG.BASE_URL}`);
  console.log(`API Key: ${CONFIG.GEMINI_API_KEY ? CONFIG.GEMINI_API_KEY.slice(0, 10) + '...' : 'Not set'}`);
  console.log(`Auth Type: ${CONFIG.AUTH_TYPE}`);
  console.log(`Debug: ${CONFIG.DEBUG}`);

  await healthCheck();

  // Uncomment to run additional tests:
  // await chatCompletion([{ role: 'user', content: 'Hello, how are you?' }]);
  // await streamChatCompletion([{ role: 'user', content: 'Write a short poem about coding.' }]);
  // await createEmbedding('The quick brown fox jumps over the lazy dog.');
}

main().catch(console.error);

File: tools/testing/connection/python-client/gemini.py
#!/usr/bin/env python3
"""
Gemini API (OpenAI Compatible) - Python Client Integration Test

Authentication: Bearer Token
Base URL: https://generativelanguage.googleapis.com/v1beta/openai
Health Endpoint: GET /models

Uses internal packages:
  - fetch_proxy_dispatcher: Environment-aware proxy configuration
  - fetch_client: HTTP client with auth support
  - provider_api_getters: API key resolution
  - app_static_config_yaml: YAML configuration loading
"""
import asyncio
import json
import os
import sys
from pathlib import Path
from typing import Any

# ============================================================================
# Project Setup - Add packages to path
# ============================================================================
PROJECT_ROOT = Path(__file__).parent.parent.parent.parent.parent
sys.path.insert(0, str(PROJECT_ROOT / "packages_py" / "app_static_config_yaml" / "src"))
sys.path.insert(0, str(PROJECT_ROOT / "packages_py" / "provider_api_getters" / "src"))
sys.path.insert(0, str(PROJECT_ROOT / "packages_py" / "fetch_client" / "src"))
sys.path.insert(0, str(PROJECT_ROOT / "packages_py" / "fetch_proxy_dispatcher" / "src"))

# Load static config
from static_config import load_yaml_config, config as static_config
CONFIG_DIR = PROJECT_ROOT / "common" / "config"
load_yaml_config(config_dir=str(CONFIG_DIR))

# Import internal packages
from fetch_proxy_dispatcher import get_proxy_dispatcher
from fetch_client import create_client_with_dispatcher, AuthConfig
from provider_api_getters import GeminiOpenAIApiToken, ProviderHealthChecker

# ============================================================================
# Configuration - Exposed for debugging
# ============================================================================
provider = GeminiOpenAIApiToken(static_config)
api_key_result = provider.get_api_key()

CONFIG = {
    # From provider_api_getters
    "GEMINI_API_KEY": api_key_result.api_key,
    "AUTH_TYPE": api_key_result.auth_type,

    # Base URL (from provider or override)
    "BASE_URL": provider.get_base_url() or "https://generativelanguage.googleapis.com/v1beta/openai",

    # SSL/TLS Configuration (runtime override, or use YAML config)
    "SSL_VERIFY": False,  # Set to None to use YAML config
    "CERT": os.getenv("CERT"),  # Client certificate path
    "CA_BUNDLE": os.getenv("CA_BUNDLE"),  # CA bundle path

    # Proxy Configuration
    "PROXY": os.getenv("HTTPS_PROXY") or os.getenv("HTTP_PROXY"),

    # Debug
    "DEBUG": os.getenv("DEBUG", "true").lower() not in ("false", "0"),
}


# ============================================================================
# Health Check
# ============================================================================
async def health_check() -> dict[str, Any]:
    """Health check using ProviderHealthChecker."""
    print("\n=== Gemini Health Check (ProviderHealthChecker) ===\n")

    checker = ProviderHealthChecker(static_config)
    result = await checker.check("gemini")

    print(f"Status: {result.status}")
    if result.latency_ms:
        print(f"Latency: {result.latency_ms:.2f}ms")
    if result.message:
        print(f"Message: {result.message}")
    if result.error:
        print(f"Error: {result.error}")

    return {"success": result.status == "connected", "result": result}


# ============================================================================
# Client Factory
# ============================================================================
def create_gemini_client():
    """Create Gemini client with standard config."""
    return create_client_with_dispatcher(
        base_url=CONFIG["BASE_URL"],
        auth=AuthConfig(type="bearer", raw_api_key=CONFIG["GEMINI_API_KEY"]),
        verify=CONFIG["SSL_VERIFY"],
        cert=CONFIG["CERT"],
        ca_bundle=CONFIG["CA_BUNDLE"],
        proxy=CONFIG["PROXY"],
    )


# ============================================================================
# Sample API Calls using fetch_client
# ============================================================================
async def chat_completion(messages: list[dict], model: str = "gemini-1.5-flash") -> dict[str, Any]:
    """Chat completion using fetch_client."""
    print(f"\n=== Chat Completion ({model}) ===\n")

    client = create_gemini_client()

    async with client:
        response = await client.post(
            "/chat/completions",
            json={"model": model, "messages": messages},
        )

        print(f"Status: {response['status']}")
        if response["ok"] and "choices" in response["data"]:
            content = response["data"]["choices"][0].get("message", {}).get("content", "")
            print(f"Response: {content}")
        else:
            print(f"Response: {json.dumps(response['data'], indent=2)}")

        return {"success": response["ok"], "data": response["data"]}


async def stream_chat_completion(messages: list[dict], model: str = "gemini-1.5-flash") -> dict[str, Any]:
    """Streaming chat completion using fetch_client SSE support."""
    print(f"\n=== Streaming Chat Completion ({model}) ===\n")

    client = create_gemini_client()

    full_content = ""
    async with client:
        async for event in client.stream(
            "/chat/completions",
            method="POST",
            json={"model": model, "messages": messages, "stream": True},
        ):
            if event.data == "[DONE]":
                continue
            try:
                data = json.loads(event.data)
                content = data.get("choices", [{}])[0].get("delta", {}).get("content", "")
                full_content += content
                print(content, end="", flush=True)
            except json.JSONDecodeError:
                pass

    print("\n")
    return {"success": True, "content": full_content}


async def create_embedding(input_text: str, model: str = "text-embedding-004") -> dict[str, Any]:
    """Create embedding using fetch_client."""
    print(f"\n=== Create Embedding ({model}) ===\n")

    client = create_gemini_client()

    async with client:
        response = await client.post(
            "/embeddings",
            json={"model": model, "input": input_text},
        )

        print(f"Status: {response['status']}")
        if response["ok"] and "data" in response["data"]:
            embedding = response["data"]["data"][0].get("embedding", [])
            print(f"Embedding dimensions: {len(embedding)}")
            print(f"First 5 values: {embedding[:5]}")
        else:
            print(f"Response: {json.dumps(response['data'], indent=2)}")

        return {"success": response["ok"], "data": response["data"]}


# ============================================================================
# Run Tests
# ============================================================================
async def main():
    """Run connection tests."""
    print("Gemini API Connection Test (Python Client Integration)")
    print("=" * 55)
    print(f"Base URL: {CONFIG['BASE_URL']}")
    print(f"API Key: {CONFIG['GEMINI_API_KEY'][:10]}..." if CONFIG['GEMINI_API_KEY'] else "API Key: Not set")
    print(f"Auth Type: {CONFIG['AUTH_TYPE']}")
    print(f"Debug: {CONFIG['DEBUG']}")

    await health_check()

    # Uncomment to run additional tests:
    # await chat_completion([{"role": "user", "content": "Hello, how are you?"}])
    # await stream_chat_completion([{"role": "user", "content": "Write a short poem about coding."}])
    # await create_embedding("The quick brown fox jumps over the lazy dog.")


if __name__ == "__main__":
    asyncio.run(main())
